pdns/ipcipher.cc

```diff
@@ -59,10 +59,40 @@ static ComboAddress encryptCA6(const ComboAddress& address, const std::string& k
 
   ComboAddress ret = address;
 
+#if OPENSSL_VERSION_MAJOR >= 3
+  auto ctx = std::unique_ptr<EVP_CIPHER_CTX, decltype(&EVP_CIPHER_CTX_free)>(EVP_CIPHER_CTX_new(), &EVP_CIPHER_CTX_free);
+  if (ctx == nullptr) {
+    throw pdns::OpenSSL::error("encryptCA6: Could not initialize cipher context");
+  }
+
+  EVP_CIPHER_CTX_set_padding(ctx.get(), 0);
+
+  auto aes128cbc = std::unique_ptr<EVP_CIPHER, decltype(&EVP_CIPHER_free)>(EVP_CIPHER_fetch(nullptr, "AES-128-CBC", nullptr), &EVP_CIPHER_free);
+
+  // NOLINTNEXTLINE(*-cast): Using OpenSSL C APIs.
+  if (EVP_EncryptInit(ctx.get(), aes128cbc.get(), reinterpret_cast<const unsigned char*>(key.c_str()), nullptr) == 0) {
+    throw pdns::OpenSSL::error("encryptCA6: Could not initialize encryption algorithm");
+  }
+
+  int len = 0;
+  // NOLINTNEXTLINE(*-cast): Using OpenSSL C APIs.
+  const auto* input = reinterpret_cast<const unsigned char*>(&address.sin6.sin6_addr.s6_addr);
+  // NOLINTNEXTLINE(*-cast): Using OpenSSL C APIs.
+  auto* output = reinterpret_cast<unsigned char*>(&ret.sin6.sin6_addr.s6_addr);
+  auto inSize = sizeof(address.sin6.sin6_addr.s6_addr);
+  if (EVP_EncryptUpdate(ctx.get(), output, &len, input, static_cast<int>(inSize)) == 0) {
+    throw pdns::OpenSSL::error("encryptCA6: Could not encrypt address");
+  }
+
+  if (EVP_EncryptFinal_ex(ctx.get(), output + len, &len) == 0) {
+    throw pdns::OpenSSL::error("encryptCA6: Could not finalize address encryption");
+  }
```

Comment:
Shouldn't be summing up the length returned by `EVP_EncryptUpdate` and the one returned by `EVP_EncryptFinal_ex`? And same for the decryption? In practice it might work because we only have one block, but that does seem brittle.

Reply:
So I'm not entirely sure what you mean. The "final" piece of encryped/decrypted data needs to be written starting at byte `output + len` (where `len` is coming from the previous call to an encryption/decryption routine, in this case `EVP_EncryptUpdate`).

I'm definitely missing something, you know better, but Isn't that also independent of how many blocks there are?

Additionally, the man page mentions the following:

```
If padding is disabled then EVP_EncryptFinal_ex() will not encrypt any more data and it will return an error if any data remains in a partial block: that is if the total data length
           is not a multiple of the block size.
```

Reply:
What I have in mind is that `EVP_EncryptUpdate` might be writing `N` bytes to `output`, and then `EVP_EncryptFinal_ex` might be adding `F` bytes to that. The current code correctly account for the `N` bytes by passing `output + len` to `EVP_EncryptFinal_ex` but then overwrites `len`, so we never know exactly how many bytes were written.
You are right that this works today, because:
- our data size is a multiple of the block size, with no padding, so  `EVP_EncryptFinal_ex` doesn't write anything
- we do not care about the final length

What I'm worried about is that someone later does a copy/paste of that code, or reuse it somewhere else. So I would prefer that we use a different variable to get the number of bytes written by  `EVP_EncryptFinal_ex` , and check that the sum of `N`+`F` is equal to `inSize`. It will always be true for this exact case, but might not if we reuse the code later for something else.

Reply:
Reasonable. I added it as an assertion.